### PR TITLE
fix(markdown): preserve inline rich text formatting on export

### PR DESCRIPTION
**Pull request by jwhite**
## Why this change is needed

AFFiNE stores paragraph text as **Y.Text delta arrays** — a sequence of \`{ insert, attributes }\` runs where \`attributes\` carries \`bold\`, \`italic\`, \`strike\`, \`code\`, and \`link\` markers. For example, the sentence \`Hello **world**\` is stored as two runs:

\`\`\`json
[
  { "insert": "Hello " },
  { "insert": "world", "attributes": { "bold": true } }
]
\`\`\`

The previous \`export_doc_markdown\` path read \`block.text\` from the serialised block, which is the **plain-text snapshot** — it strips every attribute and silently discards all inline formatting. A document full of bold, italic, and linked text would export as plain prose with no indication that any marks ever existed.

### \`src/markdown/types.ts\`
Adds an optional \`deltas?: TextDelta[]\` field to \`MarkdownRenderableBlock\` so that collected blocks can carry the raw delta array from Y.Text alongside the plain-text fallback.

### \`src/tools/docs.ts\`
- Adds \`asDeltaArray(value: Y.Text | unknown): TextDelta[]\` — reads \`value.toDelta()\` when the value is a live Y.Text instance, returns an empty array otherwise.
- In \`collectDocForMarkdown\`, populates \`block.deltas\` from \`asDeltaArray(prop.value)\` for every text block.

### \`src/markdown/render.ts\`
- Adds \`deltasToMarkdown(deltas: TextDelta[]): string\` — converts a delta array to a markdown string. It streams bold/italic markers across adjacent runs to avoid doubled markers at run boundaries (e.g. \`**bold** **text**\` → \`**bold text**\`), and wraps code spans, strikethrough, and inline links correctly.
- \`blockToMarkdown\` now prefers \`block.deltas\` when present; falls back to \`block.text\` for blocks that do not carry a delta array.

## Impact

\`export_doc_markdown\`, \`duplicate_doc\`, and \`create_doc_from_template\` all call \`collectDocForMarkdown\`, so all three now emit correct markdown for documents with inline formatting. No schema or API surface changes.

## Test plan
- [ ] Export a document that contains bold, italic, strikethrough, inline code, and a hyperlink — verify the markdown output contains the expected markers
- [ ] Export a plain-text document — verify output is identical to before
- [ ] Run \`npm run build\` — no type errors

## Summary by CodeRabbit

* **New Features**
  * Rich-text formatting (bold, italic, strikethrough, code, links) preserved in markdown export for block content and table cells via delta-based rendering.

* **Tests**
  * Added a regression suite validating markdown rich-text export round-trips and delimiter correctness.
  * Added a live integration test verifying import/export round-trip preserves rich-text marks.

* **Chores**
  * Added two markdown-related test scripts to the project scripts.

[Review Change Stack](https://app.coderabbit.ai/change-stack/DAWNCR0W/affine-mcp-server/pull/176)

### CODE REVIEW

File: src/markdown/render.ts [hunk 1]
```diff
@@ -65,6 +65,49 @@ function renderTable(tableData: string[][]): string[] {
...
+      if (wantBold && wantItalic) { inner = `***${inner}***`; }
+      else if (wantBold) { inner = `**${inner}**`; }
+      else if (wantItalic) { inner = `*${inner}*`; }
+      if (attrs.link) { inner = `[${inner.replace(/]/g, "\\]")}](${attrs.link})`; }
+      if (attrs.strike) { inner = `~~${inner}~~`; }
+      result += inner;
+      continue;
+    }
+
+    if (italicOpen && !wantItalic) { result += "*"; italicOpen = false; }
```
Comment (DAWNCR0W) [inline src/markdown/render.ts]: This state machine breaks when bold and italic partially overlap. For example, bold -> bold+italic -> italic currently emits `**A*B**C*`, which markdown-it parses as `<strong>A*B</strong>C*`, so the intended marks are not preserved. Could we make the marker transitions stack-safe or render these runs with self-contained wrappers, and add a regression test for both overlap directions?

File: src/tools/docs.ts [hunk 2]
```diff
@@ -2590,6 +2612,7 @@ export function registerDocTools(server: McpServer, gql: GraphQLClient, defaults
         sourceId: asStringOrNull(block.get("prop:sourceId")),
         caption: asStringOrNull(block.get("prop:caption")),
         tableData: block.get("sys:flavour") === "affine:table" ? extractTableData(block) : null,
+        deltas: asDeltaArray(block.get("prop:text")),
```
Comment (DAWNCR0W) [inline src/tools/docs.ts:2668]: This only attaches deltas from block `prop:text`, so table cells still go through `extractTableData()`/`richTextValueToString()` and lose inline formatting on export. Since `tableCellDeltas` already exists in the markdown types, could we carry cell deltas into the renderer as well, or explicitly narrow this PR away from table export?

**Review by DAWNCR0W — CHANGES_REQUESTED** [review]
Thanks for the PR. The direction looks good, and I confirmed the build plus the markdown round-trip test pass locally. I am requesting changes for two export-fidelity gaps that can produce incorrect Markdown or leave table rich text unpreserved. Once these are covered with focused tests, I think this should be ready for another pass.

**Comment by jwhite** [comment]
Good feedback.  I'll take a look.  I want the table export explicitly for
my use case.

Hence the PR - I'm using this now for my case and fixing issues as I find
them.

File: src/markdown/render.ts [hunk 3]
```diff
@@ -65,6 +73,26 @@ function renderTable(tableData: string[][]): string[] {
...
+  // Each run is wrapped in its own self-contained markers so that partially
+  // overlapping bold/italic sequences (e.g. bold→bold+italic→italic) always
+  // produce valid, unambiguous markdown. `_` is used for italic and `**` for
+  // bold so the two delimiter characters never collide at run boundaries.
+  let result = "";
+  for (const d of deltas) {
+    const attrs = d.attributes ?? {};
+    let inner = d.insert;
+    if (attrs.code) { result += `\`${inner}\``; continue; }
+    if (attrs.bold && attrs.italic) { inner = `**_${inner}_**`; }
```
Comment (DAWNCR0W) [inline src/markdown/render.ts]: Thanks for the follow-up here. I rechecked this against markdown-it and the current self-contained output still loses one mark when adjacent runs collide. For example, `**A****_B_**_C_` parses as `<strong>A</strong>**<em>B</em>**<em>C</em>`, so `B` is no longer bold and the literal `**` remains. Could we make the renderer produce a parser-safe sequence and pin it with a parse/round-trip assertion? Using `***...***` for bold+italic runs, with adjacent equal-attribute runs coalesced first, looks like one small direction to explore.

File: tests/test-markdown-rich-text-export.mjs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    { insert: 'B', attributes: { bold: true, italic: true } },
+    { insert: 'C', attributes: { italic: true } },
+  ];
+
+  const result = deltasToMarkdown(deltas);
+
+  // Each run is wrapped with self-contained markers; `_` is used for italic so
+  // that `**` (bold) and `_` (italic) delimiters never form ambiguous sequences.
+  assert.equal(result, '**A****_B_**_C_',
+    'bold → bold+italic → italic: self-contained per-run markers expected');
```
Comment (DAWNCR0W) [inline tests/test-markdown-rich-text-export.mjs]: Could this regression assert the parsed delta shape instead of only the emitted string? The string expectation currently passes even though markdown-it parses it with literal `**` and drops bold from the overlap segment. A focused helper that renders with `deltasToMarkdown()`, feeds the result into `parseMarkdownToOperations()`, and compares the returned deltas would catch this.

File: tests/test-markdown-rich-text-import-live.mjs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import { Client } from "@modelcontextprotocol/sdk/client/index.js";
+import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
+
+const SERVER_URL = "http://nas.fairleadsoftware.com:3002/mcp";
```
Comment (DAWNCR0W) [inline tests/test-markdown-rich-text-import-live.mjs]: This live helper looks useful locally, but I would keep hardcoded server/workspace details out of the PR. Could we either remove it from this change or make it env-driven and assertion-based, e.g. require `AFFINE_MCP_HTTP_URL`, `AFFINE_MCP_HTTP_TOKEN`, and `AFFINE_WORKSPACE_ID`, then fail when the exported marks are not present?

**Review by DAWNCR0W — COMMENTED** [review]
Thanks for the follow-up commits. The table-cell delta wiring looks like it addresses the earlier table export gap. I left a few focused comments on the remaining parser-safety issue and the live-test cleanup that would make this easier to approve with confidence.

File: src/markdown/render.ts [hunk 6]
```diff
@@ -65,6 +73,93 @@ function renderTable(tableData: string[][]): string[] {
...
+      if (attrs.bold && attrs.italic) inner = `***${inner}***`;
+      else if (attrs.bold) inner = `**${inner}**`;
+      else if (attrs.italic) inner = `_${inner}_`;
+    }
+    if (attrs.link) {
+      const label = inner.replace(/]/g, "\\]");
+      // URLs with whitespace or angle brackets aren't valid in the
+      // non-angle-bracket destination form; switch to <…> and percent-encode
+      // any inner < or > so the destination remains parseable.
+      const needsAngle = /[\s<>]/.test(attrs.link);
```
Comment (DAWNCR0W) [inline src/markdown/render.ts:151]: One edge case still seems uncovered: URLs with unbalanced parentheses, e.g. `https://example.com/foo)bar`, render as `[x](https://example.com/foo)bar)` and parse back with the link truncated to `https://example.com/foo`. This is narrow enough to handle in a follow-up if we do not want to block this PR, but I wanted to leave it explicit before merging.

**Comment by DAWNCR0W** [comment]
Closing this PR because it has not been updated for a long time. Thanks for the contribution.
